aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/internal/event/EventRecorder.java

```diff
@@ -367,43 +380,43 @@ JSONArray getBatchOfEvents(final Cursor cursor, final HashMap<Integer, Integer>
             final String errorCode = amazonServiceException.getErrorCode();
+
+            // If the error is not a retryable error, delete the events from the local database.
+            // Else if the error is a retryable error, keep the events in the local database.
             if (!isRetryable(errorCode)) {
-                addAllEventIdsToSet(eventArray, failedEventIds);
                 log.error(
-                    String.format("Failed to submit events to EventService: statusCode: " +
-                            amazonServiceException.getStatusCode() + " errorCode: ", errorCode),
+                        String.format(Locale.getDefault(), "Failed to submit events to EventService: statusCode: " +
+                                amazonServiceException.getStatusCode() + " errorCode: ", errorCode),
                         amazonServiceException);
-                log.error(String.format("Failed submission of %d events, events will be removed", eventArray.length()),
+                log.error(
+                        String.format(Locale.getDefault(), "Failed submission of %d events, events will be " +
+                                "removed from the local database. ", eventArray.length()),
                         amazonServiceException);
             } else {
                 log.error(
                         String.format("Unable to successfully deliver events to server. " +
-                                "Events will be saved, error is likely recoverable. " +
-                                "Response Status code: %s, Response Error Code: %s",
+                                        "Events will be saved, error is likely recoverable. " +
+                                        "Response Status code: %s, Response Error Code: %s",
                                 amazonServiceException.getStatusCode(), amazonServiceException.getErrorCode()),
                         amazonServiceException);
+                batchIdsAndSizeToDelete.clear();
             }
         } catch (final Exception exception) {
+            // For all other exceptions occurred during submit events such as network being
+            // offline, keep the events in the local database.
             log.error("Unable to successfully deliver events to server. " +
                     "Events will be saved, error likely recoverable." + exception.getMessage(), exception);
-        }
-        return processedEvents;
-    }
-
-    private void addAllEventIdsToSet(JSONArray eventArray, Set<String> failedEventIds) {
-        for(int i = 0; i < eventArray.length(); i++) {
-            try {
-                failedEventIds.add(eventArray.getJSONObject(i).getString(EVENT_ID));
-            } catch (JSONException e) {
-                log.error("Failed to get event id while processing event item response.", e);
-            }
+            batchIdsAndSizeToDelete.clear();
```

Comment:
We should only retry on known exceptions otherwise we may cause in infinite retry.

Reply:
Currently keeping the events in the local database only when the network is not available. For other exceptions, the exception is logged and the events will be removed from the local database.